README.rst
Fosdick python client
=====================

.. image:: https://img.shields.io/pypi/v/fosdick.svg
    :target: https://pypi.python.org/pypi/fosdick/2.0.0
.. image:: https://img.shields.io/pypi/l/fosdick.svg
    :target: https://pypi.python.org/pypi/fosdick/2.0.0

The Fosdick API Client for Python is a client library for accessing Fosdick APIs.

********
Features
********
* Get a list of inventory levels for products
* Get a list of returned orders/items
* Get a list of shipped orders
* Get a list of shipped line item
* Get a list of receipts
* Post order

************
Installation
************
Install with pip:

.. code:: python

    pip install fosdick

*************
Example Usage
*************

.. code:: python

   from fosdick.api import FosDickAPI
    
   api = FosDickAPI(<YOUR_FOSDICK_API_USERNAME>, <YOUR__FOSDICK_API_PASSWORD>)
   
    
* Get shipments

Parameters

==================  ================ ===================   =====================================================
Parameter Name        Required        Type(max length)                       Description
==================  ================ ===================   =====================================================
page                  No              numeric                    Page number of items to display.
per_page              No              numeric                    The number of items to return per page (default:
                                                                 all returns returned)
updated_at_min        No              datetime                   Only retrieve shipments updated since
updated_at_max        No              datetime                   Only retrieve shipments updated before
shipped_on_min        No              datetime                   Only retrieve shipments shipped since
shipped_on_max        No              datetime                   Only retrieve shipments shipped before
fosdick_order_num     No              varchar(17)                Search by Fosdick order number
external_order_num    No              varchar(50)                Search by External order number
==================  ================ ===================   =====================================================

.. code:: python
   
   shipments = api.get_shipments()
   
   # sample example response
    [
      {
        "fosdick_order_num": "00101201506460001",
        "external_order_num": "0011001",
        "ship_date": "2015-02-06",
        "trackings": [
          {
            "tracking_num": "9274899998944522337",
            "carrier_code": "92",
            "carrier_name": "FEDEX SMART POST"
          },
          {
            "tracking_num": "9274899998944599999",
            "carrier_code": "92",
            "carrier_name": "FEDEX SMART POST"
          }
        ]
      },
      {
        "fosdick_order_num": "00101201506460002",
        "external_order_num": "0011002",
        "ship_date": "2015-02-06",
        "trackings": [
          {
            "tracking_num": "1Z44526832337",
            "carrier_code": "3E",
            "carrier_name": "UPS GROUND"
          }
        ]
      }
    ]
	

* Get inventory

Parameters

================  ================ ===================   ===================================================
Parameter Name        Required      Type(max length)      Description
================  ================ ===================   ===================================================
  page                  No             numeric            Page number of products to display. 0 based index.
  per_page              No             numeric            The number of products to return per page
                                                          (default: all products returned)
  updated_at_min        No             datetime           Only retrieve inventory updated since
  updated_at_max        No             datetime           Only retrieve inventory updated before
================  ================ ===================   ===================================================

.. code:: python

   inventory = api.get_inventory()
   
   # example response for inventory
   [
	   {
		"sku" : "EXAMPSKU",
		"available" : true,
		"ct_quantity" : 200,
		"nv_quantity" : 0,
		"other_quantity" : 15,
		"committed" : 10,
		"available_quantity": 205,
		"updated_at":"2014-03-12T13:17:30-04:00"
	   },
	   {
		"sku" : "EXAMPSKU2",
		"available" : false,
		"ct_quantity" : 20,
		"nv_quantity" : 0,
		"other_quantity" : 0,
		"committed" : 20,
		"available_quantity": 0,
		"updated_at":"2014-03-12T13:17:30-04:00"

	   }
    ]

* Get returned orders/items

Parameters

=================  ================ ===================   ===================================================
Parameter Name        Required      Type(max length)      Description
=================  ================ ===================   ===================================================
  page                  No             numeric            Page number of items to display.
  per_page              No             numeric            The number of items to return per page (default:
                                                          all returns returned)
  updated_at_min        No             datetime           Only retrieve returns updated since
  updated_at_max        No             datetime           Only retrieve returns updated before
  returned_at_min       No             datetime           Only retrieve returns since
  returned_at_max       No             datetime           Only retrieve returns before
=================  ================ ===================   ===================================================

.. code:: python

   returned_items = api.get_all_returns()
   
   # example response for returned_items
   [
	{
		"fosdick_order_num" : "00101201456768765",
		"external_order_num" : "9912A",
		"sku" : "EXAMPSKU",
		"line_item" : 1,
		"external_line_item" : "9912A-1",
		"return_date" : "2014-03-12T9:00:30-04:00",
		"quantity_returned" : 1,
		"quality" : 0,
		"reason_code" : 2,
		"reason_description" : "Defective",
		"action_requested" : "Refund",
		"updated_at" : "2014-03-12T13:17:30-04:00"
	},
	{
		"fosdick_order_num" : "00101201456769988",
		"external_order_num" : "10012R",
		"sku" : "EXAMPSKU",
		"line_item" : 1,
		"external_line_item" : "10012R-3",
		"return_date" : "2014-03-12T9:00:30-04:00",
		"quantity_returned" : 2,
		"quality" : 1,
		"reason_code" : 5,
		"reason_description" : "Never Ordered",
		"action_requested" : "Refund",
		"updated_at" : "2014-03-12T13:17:30-04:00"
	}
   ]

* Get shipped line item

Parameters

==================  ================ ===================   ===================================================
Parameter Name        Required        Type(max length)                       Description
==================  ================ ===================   ===================================================
page                  No              numeric                    Page number of items to display.
per_page              No              numeric                    The number of items to return per page (default:
                                                                 all returns returned)
updated_at_min        No              datetime                   Only retrieve shipments updated since
updated_at_max        No              datetime                   Only retrieve shipments updated before
shipped_on_min        No              datetime                   Only retrieve shipments shipped since
shipped_on_max        No              datetime                   Only retrieve shipments shipped before
fosdick_order_num     No              varchar(17)                Search by Fosdick order number
external_order_num    No              varchar(50)                Search by External order number
==================  ================ ===================   ===================================================

.. code:: python

   shipped_detail = api.get_shipment_details()
   
   # sample response
   [
    {
            "fosdick_order_num": "00101201506460001",
            "fosdick_line_num": "1",
            "sku": "PROD001",
            "quantity": 1,
            "external_order_num": "10011001",
            "external_line_num": null,
            "external_sku": "PROD001",
            "ship_date": "2015-02-06",
            "trackings": [
              {
                "tracking_num": "9274899998944522337",
                "carrier_code": "92",
                "carrier_name": "FEDEX SMART POST"
              },
              {
                "tracking_num": "9274899998944599999",
                "carrier_code": "92",
                "carrier_name": "FEDEX SMART POST"
              }
            ]
           },
           {
            "fosdick_order_num": "00101201506460001",
            "fosdick_line_num": "2",
            "sku": "PROD002",
            "quantity": 1,
            "external_order_num": "10011001",
            "external_line_num": null,
            "external_sku": "PROD002",
            "ship_date": "2015-02-06",
            "trackings": [
              {
                "tracking_num": "1Z44526832337",
                "carrier_code": "3E",
                "carrier_name": "UPS GROUND"
              }
            ]
    }
   ]
   
* Get receipts

Parameters

==================  ================ ===================   ===================================================
page                    No              numeric                    Page number of items to display.
per_page                No              numeric                    The number of items to return per page (default:
                                                                   all returns returned)
transaction_at_min      No              datetime                   Only retrieve receipts with transaction time
                                                                   since
transaction_at_max      No              datetime                   Only retrieve receipts with transaction time
                                                                   before
updated_at_min          No              datetime                   Only retrieve receipts updated since
updated_at_max          No              datetime                   Only retrieve receipts updated before
sku                     No              varchar(50)                Search by SKU
warehouse               No              char(2)                    Search by warehouse (CT or NV)
==================  ================ ===================   ===================================================

.. code:: python

   receipts = api.get_receipts()
   
   # sample response
   [
        {
            "date_time": "2015-10-14T10:46:21-04:00",
            "warehouse": "NV",
            "receiver_num": 101263,
            "container_num": "CONTAINER-t",
            "po_num": "2101",
            "carrier_name": "UPS GROUND",
            "sku": "778888",
            "description_product": null,
            "qty": 4800,
            "num_of_floor_loaded": 60,
            "num_of_skids": 3,
            "num_of_cartons": 60,
            "updated_at": "2015-10-14T10:48:00.75-04:00"
            },
            {
            "date_time": "2015-10-13T13:24:16-04:00",
            "warehouse": "NV",
            "receiver_num": 101262,
            "container_num": null,
            "po_num": "2101",
            "carrier_name": null,
            "sku": "#B0X8",
            "description_product": null,
            "qty": 1200,
            "num_of_floor_loaded": 0,
            "num_of_skids": 2,
            "num_of_cartons": 0,
            "updated_at": "2015-10-13T13:28:07.63-04:00"
        }
    ]

* Post order

.. code:: python

    from fosdick.api import PlaceOrder

    # sample order
    order_item = {
       "UnitycartOrderPost":{
          "ClientName":"TEST",
          "ClientCode":"ad54LIADFJ2754",
          "Test":"y",
          "Order":[
             {
                "Subtotal":"0.00",
                "Total":"0.00",
                "ExternalID":"LMTB-100466",
                "AdCode":"DTC",
                "ShipFirstname":"tes1",
                "ShipLastname":"test2",
                "ShipAddress1":"test_address",
                "ShipCity":"agat",
                "ShipState":"ID",
                "ShipPhone":"",
                "ShipZip":"96915",
                "Email":"test@gmail.com",
                "UseAsBilling":"y",
                "PaymentType":"5",
                "Items":{
                   "Item":[
                      {
                         "NumOfPayments":"1",
                         "Inv":"811934020015",
                         "Qty":1,
                         "PricePer":"0.00"
                      }
                   ]
                }
             }
          ]
       }
    }

    # y to denote test order.
    test_flag = 'y'

    order = PlaceOrder(order_item, <CLIENT_CODE>, <CLIENT_NAME>, test_flag)

    # to place the order
    item = order.create_order()

    # sample response
    {
        "UnitycartOrderResponse":{
          "@xml:lang":"en-US",
          "OrderResponse":{
             "@ExternalID":"ABCD-100467",
             "SuccessCode":"True",
             "OrderNumber":"603326202469"
            }
            }
    }


*******
Support
*******

Python 2.6 and 2.7, 3.3, 3.4 & 3.5 are supported.

************
Contributors
************

- `Rajesh Krishnan P L <https://in.linkedin.com/in/rajesh-krishnan-p-l-60382a125>`_
- `Abdul Niyas P M <https://github.com/abdulniyaspm1>`_

*******
License
*******
MIT


CHANGELOG.rst
##########
Changelog
##########


****************
2.0.0 (2017-02-01)
****************

* First working version

fosdick/api.py
import json
import logging
import requests

try:
    # python 2
    from urlparse import urljoin
except ImportError:
    # python3
    from urllib.parse import urljoin

try:
    from collections import OrderedDict
except ImportError:
    # python 2.6 or earlier, use backport
    from ordereddict import OrderedDict

from .config import *


class FosDickAPIException(Exception):
    pass


class FosDickAPIUnavailable(FosDickAPIException):
    def __init__(self, url, message=None):
        self.message = message or "API is unavailable"
        self.url = url

    def __str__(self):
        return self.message


class FosDickAPIRequestError(FosDickAPIException):
    def __init__(self, message):
        super(FosDickAPIRequestError, self).__init__(message)


class FosDickAPI(object):
    """
    This class will interact with fostdick api's for pulling information
    about shipments, inventory, receipts etc.
    """
    # sets "init'ial" state .
    def __init__(self, username=None, password=None):
        self.USERNAME = username or FOSDICK_API_USERNAME
        self.PASSWORD = password or FOSDICK_API_PASSWORD
        self.URL = URL

    def _get_url(self, pathname):
        """
        For each pathname return corresponding URL.
        :param pathname:
        :return:
        """

        return urljoin(self.URL, pathname)

    def _create_request(self, action, **kwargs):
        """
        Used to create the request_url and query params.
        :param action: action
        :param kwargs:
        :return: request_url, data
        """
        data = dict()
        data.update(kwargs)
        req_url = URL_MAP.get(action, '/')
        request_url = self._get_url(req_url)
        return request_url, data

    def _get_response(self, request_url, data):
        """
        Perform json-request to API
        :param request_url: Absolute url to API.
        :type request_url: str
        :param data: Datas to send to the API.
        :type data: dict
        :return: json response
        :rtype: dict
        """
        # To authenticate the request.
        auth = (self.USERNAME, self.PASSWORD)
        try:
            req = requests.get(
                request_url,
                params=data,
                auth=auth
            )
            content = req.content
            logging.debug(content)
        except requests.exceptions.RequestException as e:
            # API not available.
            raise FosDickAPIUnavailable(self.URL, str(e))
        try:
            json_response = req.json()
        except Exception as e:
            raise FosDickAPIRequestError(str(e))
        return json_response

    def request(self, action, **kwargs):
        """
        Perform json-request to API
        :param action: method action (name of the function in API)
        :type action: str
        :return: json response
        :rtype: dict
        """
        request_url, data = self._create_request(
            action, **kwargs)
        try:
            json_response = self._get_response(request_url, data)
        except Exception as e:
            raise FosDickAPIException(str(e))
        # if No response
        if not json_response:
            return []

        return json_response

    def get_shipments(self, **kwargs):
        """
        :param kwargs:page,
                per_page,
                updated_at_min,
                updated_at_max,
                shipped_on_min,
                shipped_on_max,
                fosdick_order_num,
                external_order_num
        :return: list of shipped orders
        """
        shipment_info = self.request(action='shipments', **kwargs)
        return shipment_info

    def get_inventory(self, **kwargs):
        """
        :param kwargs:page,
                per_page,
                updated_at_min,
                updated_at_max.
        :return: List of inventory levels for products
        """
        inventory_info = self.request(action='inventory', **kwargs)
        return inventory_info

    def get_all_returns(self, **kwargs):
        """
        :param kwargs:page,
                    per_page,
                    updated_at_min,
                    updated_at_max,
                    returned_at_min,
                    returned_at_max.
        :return:List of returned orders/items
        """
        returns_items = self.request(action='returns', **kwargs)
        return returns_items

    def get_shipment_details(self, **kwargs):
        """
        :param kwargs:page,
                per_page,
                updated_at_min,
                updated_at_max,
                shipped_on_min,
                shipped_on_max,
                fosdick_order_num,
                external_order_num
        :return:List of shipped line item.
        """
        shipment_details = self.request(action='shipmentdetail', **kwargs)
        return shipment_details

    def get_receipts(self, **kwargs):
        """
        :param kwargs: page,
                per_page,
                transaction_at_min,
                transaction_at_max,
                updated_at_min,
                updated_at_max,
                sku,
                warehouse
        :return: Submit a list of receipts.
        """
        receipts = self.request(action='receipts', **kwargs)
        return receipts

    # TODO: Xml request, and parsing.


class PlaceOrder(object):
    """
    This is where the order is placed"
    """
    # sets "init'ial" state .
    def __init__(self, order, client_code=None, client_name=None, test_flag=None):
        self.ORDER = order
        self.URL = PLACE_ORDER_URL
        self.CLIENT_CODE = client_code or CLIENT_CODE
        self.CLIENT_NAME = client_name or CLIENT_NAME
        self.TESTFLAG = test_flag or TESTFLAG
        self.headers = {'content-type': 'application/json'}

    def __new__(cls, order, client_code=None, client_name=None, test_flag=None):
        """
        to validate the order dic.
        """
        if not isinstance(order, dict):
            raise FosDickAPIException("Invalid order format.")
        return super(PlaceOrder, cls).__new__(cls)

    def _create_request(self):
        """
        Will return absolute url and formatted order data.
        :return:req_url, data
        """
        data = self._create_order_data()
        req_url = self.URL
        return req_url, data

    def _get_response(self, request_url, data):
        """
        Perform json-request to API.
        :param request_url: Absolute url to API.
        :type request_url: str
        :param data: Data to post to the API.
        :type data: dict
        :return: json response
        :rtype: dict
        """
        try:
            req = requests.post(
                request_url,
                data,
                headers=self.headers
            )
            content = req.content
            logging.debug(content)
        except requests.exceptions.RequestException as e:
            # API not available.
            raise FosDickAPIUnavailable(self.URL, str(e))
        try:
            json_response = req.json()
        except Exception as e:
            raise FosDickAPIRequestError(str(e))
        if json_response['UnitycartOrderResponse']['OrderResponse']['SuccessCode'] == 'False':
            raise FosDickAPIException(json_response)
        return json_response

    def _create_order_data(self):
        """
        organise order data according to requirement.
        And also order of dic items is important.
        """
        data = OrderedDict()
        data["UnitycartOrderPost"] = OrderedDict()
        data["UnitycartOrderPost"]["ClientName"] = self.CLIENT_NAME
        data["UnitycartOrderPost"]["ClientCode"] = self.CLIENT_CODE
        data["UnitycartOrderPost"]["Test"] = self.TESTFLAG
        data["UnitycartOrderPost"]["Order"] = []
        order_object = self.ORDER
        items = {"Item": []}
        order = order_object['Order'][0]
        all_items = order_object['Order'][0]['Items']['Item']
        for order_item in all_items:
            items["Item"].append(OrderedDict({
                "Inv": order_item['Inv'],
                "Qty": order_item['Qty'],
                "PricePer": order_item['PricePer'],
                "NumOfPayments": order_item['NumOfPayments'],
            }))
        order_with_items = OrderedDict()
        order_with_items["Subtotal"] = order['Subtotal']
        order_with_items["Total"] = order['Total']
        order_with_items["ExternalID"] = str(order['ExternalID'])[:60]
        order_with_items["AdCode"] = order['AdCode']
        order_with_items["ShipFirstname"] = order['ShipFirstname'][:22]
        order_with_items["ShipLastname"] = order['ShipLastname'][:16]
        order_with_items["ShipAddress1"] = order['ShipAddress1'][:26]  # total limit 30
        order_with_items["ShipCity"] = order['ShipCity'][:13]
        order_with_items["ShipState"] = order['ShipState'][:2]
        order_with_items["ShipPhone"] = ""
        order_with_items["ShipZip"] = order['ShipZip'][:11]
        order_with_items["Email"] = order['Email'][:100]  # configurable in fosdick
        order_with_items["UseAsBilling"] = order['UseAsBilling']
        order_with_items["PaymentType"] = order['PaymentType']
        order_with_items["Items"] = items
        data["UnitycartOrderPost"]["Order"].append(order_with_items)
        data = json.dumps(data)
        return data

    def request(self):
        """
        Perform json-post-request to API
        :return: json response
        :rtype: dict
        """
        request_url, data = self._create_request()
        try:
            json_response = self._get_response(request_url, data)
        except Exception as e:
            raise FosDickAPIException(str(e))
        # if No response
        if not json_response:
            return []

        return json_response

    def create_order(self):
        """
        Used to place the order.
        :return: order_item
        """
        order_item = self.request()
        return order_item

    # TODO: multiple order at a time.


